docs_src/design/adr/013-Device-Service-Events-Message-Bus.md

```diff
@@ -0,0 +1,215 @@
+
+The C SDK will also will recreate this Message Envelope.
+
+### Redis Streams
+
+The Redis Streams Message Bus implementation currently only supports password authentication in configuration. There is support in the code for certificates, but it is not exposed via configuration. The implementation may need to be enhanced to include the authentication capabilities described in the **[Secure Connections](#secure-connections)** section below.
+
+### Application Services
+
+As part of the V2 API consumption work in Ireland the App Services SDK will be changed to expect to receive V2 Event DTOs rather than the V1 Event model. It will also be updated to no longer expect or use the `Checksum` currently on the  Message Envelope. Note these changes must occur for the V2 consumption and are not directly tied to this effort. 
+
+The App Service SDK will be enhanced for the secure Message Bus connection described below. See **[Secure Connections](#secure-connections)** for details
+
+### Message Bus Topics
+
+> *Note: The change recommended here is not required for this design, but it provides a good opportunity to adopt it.*
+
+Currently Core Data publishes Events to the simple `events` topic. All Application Services running receive all Events published, whether they want them or not. The Events can be filtered out using the `FilterByDeviceName` pipeline function, but the Application Services still receive all the Events and process all the Events to some extent. This could cause load issues in a deployment with many devices and large volume of Events.
+
+Pub/Sub systems have advanced topic schema, which we can take advantage of to filter for just the Events the Application Service actual needs. If the publishers of Events add the `Device Name` to the topic in the form `edgex/events/<device-name>` then the Application Service can filter for just the Events from the device(s) it wants by only subscribing to those `Device Names`, i.e. `edgex/events/Random-Integer-Device` . If persistence is require, Core Data will subscribe using the `#` wild card, i.e. `edgex/events/#` , so that it receives all Events. 
+
+The Message Bus abstraction allows for multiple subscriptions, so an Application Service could specify to receive data from only specific devices by creating multiple subscriptions. i.e.  `edgex/Events/Random-Integer-Device` and  `edgex/Events/Random-Boolean-Device`. Currently the App SDK only allows for a single subscription topic to be configured, but that could easily be expanded to handle a list of subscriptions. See [Configuration](#configuration) section below for details. 
```

Comment:
some thoughts/questions about this... I generally like the idea but have we thought through some of these:
- for an edge platform, would the cost of an extra set of topic connections (pub and sub) be better than parsing the message for the device coming in from one topic?
- for core and app services - they would have to be configured and bootstrapped to work with an undetermined number of topics.  And what happens if a new device is added after already up?  Or if a device was removed?  Would topic per device be something you have to know that the device list is static?
- If we just deal with one topic right now, could we add this functionality in a future release as I don't think this would be non-backward compatible?  Start small and work our way to the more complex?

Reply:
> some thoughts/questions about this... I generally like the idea but have we thought through some of these:

Still only a single connection to the  Message Bus. Yes, much better since the brokers are designed to handle many different subscriptions.
Core would use the # wild card syntax to receive all events with one subscription. Yes, the subscribe topic with device name would be static, i.e. you know ahead of time the devices.
Yes, I think so, but as we are already making changes to these services it is small lift to do it now.
Note that with this in place, the current single topic could still be configured and used as it is today.
I am now wondering if this should be DeviceProfile rather than DeviceName since one would want values from all instances of device rather than an single instance.